src/main/java/build/buildfarm/server/ByteStreamService.java

```diff
@@ -262,9 +262,11 @@ private void writeOperationStream(
         String operationStream = parseOperationStream(resourceName);
 
         OutputStream outputStream = instance.getStreamOutput(operationStream);
-        outputStream.write(request.getData().toByteArray());
-        if (request.getFinishWrite()) {
-          outputStream.close();
+        if (outputStream != null) {
+          request.getData().writeTo(outputStream);
+          if (request.getFinishWrite()) {
+            outputStream.close();
```

Comment:
What's your strategy for making sure all output streams are closed? In the remote worker, we've discovered at least one file handle leak due to streams not being closed properly in all cases.